sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/IdempotentAmqpEventBatch.cs

```diff
@@ -0,0 +1,206 @@
+        ///
+        public IdempotentAmqpEventBatch(AmqpMessageConverter messageConverter,
+                                        CreateBatchOptions options,
+                                        TransportProducerFeatures activeFeatures) : this(messageConverter, options, activeFeatures, new List<EventData>())
+        {
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="AmqpEventBatch"/> class.
+        /// </summary>
+        ///
+        /// <param name="messageConverter">The converter to use for translating <see cref="EventData" /> into the corresponding AMQP message.</param>
+        /// <param name="options">The set of options to apply to the batch.</param>
+        /// <param name="activeFeatures">The flags specifying the set of special transport features have been opted-into.</param>
+        /// <param name="eventBackingStore">The backing store for holding events in the batch.</param>
+        ///
+        protected IdempotentAmqpEventBatch(AmqpMessageConverter messageConverter,
+                                           CreateBatchOptions options,
+                                           TransportProducerFeatures activeFeatures,
+                                           IList<EventData> eventBackingStore) : base(messageConverter, options, activeFeatures)
+        {
+            _sizeBytes = ReservedOverheadBytes;
+            BatchEvents = eventBackingStore;
+        }
+
+        /// <summary>
+        ///   Attempts to add an event to the batch, ensuring that the size
+        ///   of the batch does not exceed its maximum.
+        /// </summary>
+        ///
+        /// <param name="eventData">The event to attempt to add to the batch.</param>
+        ///
+        /// <returns><c>true</c> if the event was added; otherwise, <c>false</c>.</returns>
+        ///
+        public override bool TryAdd(EventData eventData)
+        {
+            Argument.AssertNotNull(eventData, nameof(eventData));
+            Argument.AssertNotDisposed(_disposed, nameof(EventDataBatch));
+
+            // Reserve space for producer-owned fields that correspond to idempotent publishing
+            // metadata needed by the service.
+
+            eventData.PendingPublishSequenceNumber = int.MaxValue;
+            eventData.PendingProducerGroupId = long.MaxValue;
+            eventData.PendingProducerOwnerLevel = short.MaxValue;
+
+            try
+            {
+                var size = _sizeBytes + MeasureBytes(eventData);
+
+                if (size > MaximumSizeInBytes)
+                {
+                    return false;
+                }
+
+                // Clone the event when accepting into the batch to ensure
+                // that its size does change if the original reference is
+                // mutated.
+
+                _sizeBytes = size;
+                BatchEvents.Add(eventData.Clone());
```

Comment:
See my previous comment about shallow vs deep clone

Reply:
The intent, as you're likely aware, is to prevent mutating the batch size by manipulating the `EventData` source.   

Generally, we're trying to guard against setting the direct properties or adding additional members into the `ApplicationProperties` dictionary.   The properties are comprised of primitives and immutable types that we copy over to the clone.  The dictionary only supports primitive types and streams for serialization - the latter of which is a very unusual case.   _([ref](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs#L942))_

In the case where someone put a stream in the application properties dictionary and is writing to it, and has idempotent retries enabled ... our abstraction leaks and manipulating the stream does mutate the batch.   I'd call it more like a mid-level copy - not fully deep but also not totally shallow.   It's not ideal, but it's been a decent balance for the mainstream case.